src/formats/apple-notes/convert-table.ts

```diff
@@ -0,0 +1,125 @@
+import { 
+	ANMergeableDataObject, ANMergeableObjectEntry,
+	ANTableKey, ANTableType
+} from './models';
+import { NoteConverter } from './convert-note';
+import { AppleNotesImporter } from '../apple-notes';
+
+export class TableConverter {
+	importer: AppleNotesImporter;
+	table: ANMergeableDataObject;
+	
+	keys: string[];
+	types: string[];
+	uuids: string[];
+	objects: ANMergeableObjectEntry[];
+	
+	rowCount: number;
+	rowIndices: Record<number, number> = {};
+	
+	columnCount: number;
+	columnIndices: Record<number, number> = {};
+	
+	constructor(importer: AppleNotesImporter, table: ANMergeableDataObject) {
+		this.importer = importer;
+		this.table = table;
+		
+		const data = table.mergeableDataObjectData;
+		
+		this.keys = data.mergeableDataObjectKeyItem;
+		this.types = data.mergeableDataObjectTypeItem;
+		this.uuids = data.mergeableDataObjectUuidItem.map(this.uuidToString);
+		this.objects = data.mergeableDataObjectEntry;
+	}
+	
+	async parse(): Promise<string[][] | null> {		
+		const root = this.objects.find(e => e.customMap && this.types[e.customMap.type] == ANTableType.ICTable);
+		if (!root) return null;
+	
+		let cellColumns: ANMergeableObjectEntry | null = null;
+		
+		for (const entry of root.customMap.mapEntry) {
+			const object = this.objects[entry.value.objectIndex];
+			
+			switch (this.keys[entry.key]) {
+				case ANTableKey.Rows:
+					this.findIndices(object, true);
+					break;
+				case ANTableKey.Columns:
+					this.findIndices(object, false);
+					break;
+				case ANTableKey.CellColumns:
+					cellColumns = object;
+					break;
+			}
+		}
+		
+		if (this.columnCount < 1 || this.rowCount < 1 || !cellColumns) return null;
+		return await this.computeCells(cellColumns!);
+	}
+	
+	findIndices(object: ANMergeableObjectEntry, rows: boolean): void {
```

Comment:
I'm having trouble following what's happening in this function without an understanding of the data model. Would it be possible to add a brief description to a couple of the methods in this class?

Reply:
I've tried to make it easier to follow by restructuring a lot of the code and adding a bunch of comments, but it's inherently not a very easy thing to understand.